### ISSUE TITLE
support for sparse arrays

### ISSUE THREAD

**Issue by djarecka** [fsspec/kerchunk#301]
I've been testing `kerchunk` and follow the [tutorial](https://fsspec.github.io/kerchunk/tutorial.html#single-file-jsons) to work with `h5ad` files. Unfortunately I've realized that the `json` reference file that I created with `kerchunk` will not work with `xr.open_dataset`, I believe this is because the original file contains sparse arrays. 

Do you have any suggestions how can I use `kerchunk` and the reference file for this kind of data?

Below the code that can reproduce my issue:
```
# creating the data and saving as a h5ad file
import numpy as np
from scipy.sparse import csr_matrix
import anndata

row = np.array([0, 0, 1, 2, 2, 2])
col = np.array([0, 2, 2, 0, 1, 2])
data = np.array([1, 2, 3, 4, 5, 6])
data_csr = csr_matrix((data, (row, col)), shape=(3, 3))

small_data = anndata.AnnData(X=data_csr)
small_data_path = '/home/jovyan/kerchunk_tutorial/small_test.h5ad'
small_data.write_h5ad(small_data_path)


# creating the json ref file with kerchunk
from kerchunk.hdf import SingleHdf5ToZarr
import fsspec
import ujson

fs_loc = fsspec.filesystem('')
so = dict(mode='rb', anon=True, default_fill_cache=False, default_cache_type='first') # args to fs.open()
outf_test = "/home/jovyan/kerchunk_tutorial/small_test_chunk.json"
with fs_loc.open(small_data_path, **so) as infile:
    h5chunks = SingleHdf5ToZarr(infile, inline_threshold=300)
    with fs_loc.open(outf_test, 'wb') as f:
        f.write(ujson.dumps(h5chunks.translate()).encode());


# using the json ref file with xarray
import xarray as xr
ds = xr.open_dataset("reference://", engine="zarr", backend_kwargs={
                    "consolidated": False,
                    "storage_options": {"fo": outf_test}
                    })
```

I don't get any error, but I get an empty dataset:
```
<xarray.Dataset>
Dimensions:  ()
Data variables:
    *empty*
```

**Comment by martindurant** [comment]
@jim22k , do you happen to know how HDF5 stores sparse data like this?

@djarecka , kerchunk's HDF5 current default behaviour is to warn and skip over any variables whose encoding it doesn't know how to translate to zarr. I'm not sure the zarr API can be made to return sparse matrices at all (it's very numpy-based), but there is a separate effort to standardise sparse storage via a zarr convention (@jim22k , again), which would probably look similar to whatever HDF5 does.

**Comment by jim22k** [comment]
The effort to standardize a binary sparse storage format is being worked [here](https://github.com/GraphBLAS/binsparse-specification). The current plan is to use HDF5 for the container, but zarr is also under consideration. The storage will use several dense arrays (pointers, indices, values for CSR) plus JSON metadata about the format. Other common sparse storage formats will also be supported.

@djarecka given that your example is using AnnData, you may want to reach out to @ivirshup who is involved in both the binsparse effort and anndata.

**Comment by ivirshup** [comment]
Hey @djarecka, I've been looking into using kerchunk for this very case! I'd really like to have this as core AnnData functionality this year. But, idk if it's possible right this moment. 

At the moment my approach for this looks like: loading the `indptr` array into memory, calculating new combined offsets, then creating a new "kerchunked" group with this indptr and kerchunk-concatenated indices and data arrays. However, I've run into some issues (https://github.com/fsspec/kerchunk/issues/303), so it's not currently working.

<details>
<summary>functions that aren't working aka "pseudocode"</summary>
```python
def concatenate_zarr_arrays(arrays: list[zarr.Array]) -> dict:
    references = [single_zarr(array.store.dir_path(array.path)) for array in arrays]
    return concatenate_arrays(references)

def concatenate_zarr_csr_arrays(arrays: list[zarr.Group]) -> dict:
    # Group metadata
    shapes = [array.attrs["shape"] for array in arrays]
    merged_shape = [sum(s[0] for s in shapes), shapes[0][1]]
    metadata = {"encoding-type": "csr_matrix", "encoding-version": "0.1.0", "shape": merged_shape}
    
    # Concatenating indices & data
    references = [single_zarr(array.store.dir_path(array.path)) for array in arrays]
    data_refs = concatenate_arrays(references, path="data")
    indices_refs = concatenate_arrays(references, path="indices")
    
    # Concatenating indptr in memory
    indptrs = [array["indptr"][:] for array in arrays]

    new_indptrs = []
    pos = indptrs[0][-1]
    for indptr in indptrs[1:]:
        new_indptr = indptr[1:]
        new_indptr += pos
        new_indptrs.append(new_indptr)
        pos = new_indptr[-1]
    new_indptr = np.concatenate([indptrs[0]] + new_indptrs)
    new_indptr_zarr = zarr.array(new_indptr)
    
    indptr_refs = {
        "version": 1,
        "refs": {f"indptr/{k}": v for k, v in new_indptr_zarr.store.items()}
    }
#     indptr_refs["refs"]["indptr/.zarray"] = indptr_refs["refs"]["indptr/.zarray"].decode()
#     for k in indptr_refs["refs"]:
#         if k.startswith("indptr/."):
#             continue
#         indptr_refs["refs"][k] = b"base64:" + indptr_refs["refs"][k]
    
    # Merging into group
    merged = merge_vars([data_refs, indices_refs, indptr_refs])
    # Setting .zattrs
    merged[".zattrs"] = json.dumps(metadata)
    
    return merged, new_indptr_zarr
```

</details>

**Comment by ivirshup** [comment]
For a little more context: [here is the hdf5 representation of sparse arrays being used here](https://anndata.readthedocs.io/en/latest/fileformat-prose.html#sparse-arrays). It's not native hdf5, it's a "convention". But it's very similar to what in the binsparse specification now.

And my current understanding is that this won't be possible with kerchunk until/ (unless) [ZEP 3](https://github.com/orgs/zarr-developers/discussions/52) is accepted and implemented.

**Comment by djarecka** [comment]
@martindurant @jim22k @ivirshup - thank you so much for your answers and suggestions! 
I haven't used sparse data before, so I also have to read more. Seems like for cell taxonomy and matrix cell vs genes it is the best option.

@ivirshup - I will try your suggestions, thank you!

**Comment by martindurant** [comment]
Out of interest, did you get any warnings while producing the kerchunked empty dataset? You can pass `error="pdb"` to SingleHdf5ToZarr to be dropped into the debugger at the point of failure.

**Comment by djarecka** [comment]
@martindurant - I didn't get any error or warning from `SingleHdf5ToZarr`, it produced the json file. `xarray` also didn't complain, just gave me en empty dataset.
It's actually interesting also because some of the data was not a sparse matrix (in my original file that is not public, not the one that I created for the issue), but perhaps it was skipped because it was saved in `data.obs` and `data.obsm` (i think this is specific to `h5ad` format, see [here](https://anndata.readthedocs.io/en/latest/tutorials/notebooks/getting-started.html#Observation/Variable-level))

**Comment by martindurant** [comment]
Right, I would need to do some reading on h5ad - kerchunk.hdf certainly doesn't scan multiple files.

**Comment by djarecka** [comment]
this is actually in one file

**Comment by martindurant** [comment]
Out of interest, what does xarray show when opening the HDF5 directly? (please redact anything sensitive)

**Comment by djarecka** [comment]
it also doesn't work... it doesn't give me an error, but just returns an empty dataset.

**Comment by martindurant** [comment]
Ah, so it seems the HDF5 is not netCDF compliant, and xarray is not the tool you would use to analyse it. However, maybe then kerchunk was doing the right thing? You can probably load the data directly with zarr. What would you normally do with it?

**Comment by djarecka** [comment]
@martindurant - perhaps it did the right thing... It's the first time I work with this type of file, so I'm just exploring the options right now, and someone mentioned your tool, but seems like this is a very specific hdf5 file and many hdf5-compatible tools might not work.
